### DISCUSSION TITLE
Interface to AS608 fingerprint sensor

### DISCUSSION THREAD

**Discussion by ggodart** [arendst/Tasmota#17823] [Q&A]
Beyond creating my own build with;
`
#define USE_AS608                          // Add support for AS608 optical and R503 capacitive fingerprint
#define USE_AS608_MESSAGES     // Add verbose error messages 
`
And configuring the TX and RX of the device, are there any documents on how to actually use it, e.g. how to enrol new fingerprints.
THX

**Comment by pkkrusty** [comment]
According to the driver, there are three commands: FPEnroll <x>, FPDelete <x>, and FPCount.  So you’ll have to keep track of who is in what slot if you’re going to be enrolling and deleting multiple users. There are 128 slots, so for the first user you would give the command `FPEnroll 1` and the console will print the commands to follow (place finger, remove finger, re-place same finger). Not sure how the timing works for it, I’m just looking at the code. Once the fingerprint is registered, I suppose you can just place your finger on it, it will decide success or failure, and publish JSON accordingly. Then up to you how to use it.

**Comment by Noschvie** [comment]
Hello, want to connect the AS608 fingerprint sensor to a ESP32 D1mini.
Are GPIO16 and GPIO17 supported for that or GPIO09 and GPIO10.
Found different pinouts.
Thanks, Norbert

**Reply by sfromis** [reply]
Well, with ESP32, there is no real question of what's "supported", as long as you have the pins available. This is normally silk screen printed on the PCB, but just select two pins, as any pin can be "hardware serial".

**Comment by Oekel** [comment]
Sorry for late response/request (but this topic matches perfect)
Are there any limitations for ESP2866?

EDIT: ...fingerprint sensor (+3k4 code) --> will check

ESP Chip ID | 4714224 (ESP8266EX)
-- | --
0x14605E (DOUT)
Flash-Chip | 1024 KB
Usable Flash-Chip | 1024 KB
Programmable | 629 KB
Free Flash | 372 KB
Free RAM | 23,0 KB

--> Should be possible! (?)

**Reply by sfromis** [reply]
Well, apart from the general lower capacity (less local processing features) with ESP8266, there is nothing precluding the basics of AS608 support working the same, with a suitable build to include the driver.

**Reply by Noschvie** [reply]
> Are there any limitations for ESP2866?

What speaks against using an ESP32? This provides more resources.

**Reply by Oekel** [reply]
I have a good self-build hardware design within the little one. Why should I switch to ESP32 when software will work and I only need GPIOs + this fingerprint?

**Reply by sfromis** [reply]
I'd see it from the other side.... Why would I bother with ESP8266 when I get better features with ESP32?

But sure, for basic operation, it is fine to use ESP8266, when you already have an old hardware design.

**Comment by Oekel** [comment]
Ok, I got the basics working on my ESP8288.

But are there realy only these **three** commands: **FPEnroll** , **FPDelete** , and **FPCount** available?
What about the two LEDs the device has? (I have RED+BLUE) Booting with red and reacting to any finger action with Purple (so I think both colours) In Case I got the confidence with the driver I have so decide by my own if the finger is correct? If so, I allso want to decide which LED is aktive for feedback.

**Comment by Noschvie** [comment]
Have a look to this
```
#define AS608_COLOR_SCAN      4    // Red = 1, Blue = 2, Purple = 3, Green = 4, Yellow = 5, Cyan = 6, White = 7
#define AS608_DUPLICATE       8    // Number of 0.25 Sec to disable detection
#endif
```

**Reply by sfromis** [reply]
There is also:
```
#define AS608_COLOR_INIT      1    // Red = 1, Blue = 2, Purple = 3, Green = 4, Yellow = 5, Cyan = 6, White = 7
```

**Reply by Oekel** [reply]
> Have a look to this
> 
> ```
> #define AS608_COLOR_SCAN      4    // Red = 1, Blue = 2, Purple = 3, Green = 4, Yellow = 5, Cyan = 6, White = 7
> #define AS608_DUPLICATE       8    // Number of 0.25 Sec to disable detection
> #endif
> ```

For sure I see a LED while SCAN and a LED while INIT.
But I want to send a LED for SUCCESS or DENIED by hand.

**Reply by kleinegitty** [reply]
Did you find out already? How can I set a red led for denied and a green led for approved?
Are there more AS608_COLOR settings available?

**Comment by Noschvie** [comment]
A rule like this can be used  `on FPrint#Id=1 do power1 on endon`

**Comment by bananajoe86** [comment]
Maybe someone here would like to modify the fingerprint. In my opinion, the LED ring around the sensor should be able to change its color using a rule. Like this. 
`void AS608_ProcessCommand(char *param) {
  if (!As608Finger) return;

  if (!strcmp(param, "red")) {
    As608Finger->LEDcontrol(FINGERPRINT_LED_ON, 0, FINGERPRINT_LED_RED);
  } else if (!strcmp(param, "blue")) {
    As608Finger->LEDcontrol(FINGERPRINT_LED_ON, 0, FINGERPRINT_LED_BLUE);
  } else if (!strcmp(param, "green")) {
    As608Finger->LEDcontrol(FINGERPRINT_LED_ON, 0, FINGERPRINT_LED_GREEN);
  } else if (!strcmp(param, "off")) {
    As608Finger->LEDcontrol(FINGERPRINT_LED_OFF, 0, 0);
  }
}`
Unfortunately, I have little knowledge of compiling Tasmota and whether the code could actually work, but perhaps someone would like to help.
